primitives/transaction-pool/src/pool.rs

```diff
@@ -217,11 +219,29 @@ pub trait TransactionPool: Send + Sync {
 	fn ready_transaction(&self, hash: &TxHash<Self>) -> Option<Arc<Self::InPoolTransaction>>;
 }
 
+/// Events that the transaction pool listens for.
+pub enum ChainEvent<TP: TransactionPool> {
+	/// New blocks have been added to the chain
+	Canonical {
+		/// Id of the just imported block.
+		id: BlockId<TP::Block>,
+		/// Header of the just imported block
+		header: <TP::Block as BlockT>::Header,
+		/// List of retracted blocks ordered by block number.
+		retracted: Vec<BlockHash<TP>>,
+	},
+	/// An existing block has been finalzied.
+	Finalized {
+		/// Hash of just finalized block
+		hash: BlockHash<TP>,
+	}
+}
+
 /// Trait for transaction pool maintenance.
-pub trait MaintainedTransactionPool : TransactionPool {
+pub trait MaintainedTransactionPool: TransactionPool {
 	/// Perform maintenance
-	fn maintain(&self, block: &BlockId<Self::Block>, retracted: &[BlockHash<Self>])
-		-> Pin<Box<dyn Future<Output=()> + Send>>;
+	fn maintain(&self, event: &ChainEvent<Self>) -> Pin<Box<dyn Future<Output=()> + Send>>
+		where Self: Sized;
```

Comment:
Why do you need `Self: Sized`?

Reply:
You get a compiler error without it. I think because `ChainEvent<Self>` may potentially be a recursive type, adding that the trait has a known size at compile time prevents this.